### PR TITLE
Add interactive masking tool

### PR DESCRIPTION
**Pull request by nvaytet**
This is https://github.com/scipp/essreduce/pull/30 moved to Scippneutron.

Part of https://github.com/scipp/essdiffraction/issues/50

- Masks can be drawn on 1d and 2d data (using `mpltoolbox`).
- They can then be saved to file, using Scipp's hdf5 format.
- Only rectangular and horizontal/vertical spans are supported for now.

Screenshot at 2024-05-24 00-06-34
Screenshot at 2024-05-24 00-05-51

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -87,6 +88,7 @@ filterwarnings = [
   'ignore:Widget.widget_types is deprecated:DeprecationWarning',
   'ignore:Widget.widgets is deprecated:DeprecationWarning',
   'ignore:`load_nexus` is deprecated:UserWarning',
+  'ignore:Passing unrecognized arguments to super:DeprecationWarning',
```
Comment (jl-wynen) [inline pyproject.toml:91]: Where does this come from?

File: requirements/basetest.in [hunk 2]
```diff
@@ -2,6 +2,7 @@
 # Do not make an environment from this file, use test.txt instead!
 
 hypothesis
+ipympl
```
Comment (jl-wynen) [inline requirements/basetest.in:5]: Could you use something like `plopp[interactive]` instead?

File: tests/conftest.py [hunk 3]
```diff
@@ -14,3 +15,8 @@ def pytest_assertrepr_compare(op: str, left: Any, right: Any) -> list[str]:
         return [f'Unit({left}) {op} Unit({right})']
     if isinstance(left, sc.DType) or isinstance(right, sc.DType):
         return [f'{left!r} {op} {right!r}']
+
+
+@pytest.fixture()
+def _use_ipympl():
+    matplotlib.use('module://ipympl.backend_nbagg')
```
Comment (jl-wynen) [inline tests/conftest.py:22]: Does this only apply to a single test? It looks like it switches to the ipympl backend for all subsequent tests.

File: tests/masking_tool_test.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+
+import json
+import tempfile
```
Comment (jl-wynen) [inline tests/masking_tool_test.py:5]: Please use `pytest.tmp_path` https://docs.pytest.org/en/6.2.x/tmpdir.html#the-tmp-path-fixture instead. This makes sure the files get cleaned up properly but still allows you to inspect them after the test ran.

File: tests/masking_tool_test.py
Comment (jl-wynen) [inline tests/masking_tool_test.py:1]: Does it make sense to make the click API public? It is jot great to rely on implementation details in tests. And maybe, this could be useful for building UIs in the future where we may have some initial masks. But this may be too hypothetical...

File: tests/masking_tool_test.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    v.value = True
+    v._tool.click(-20.0, 0.0)
+    v._tool.click(1.0, 0.0)
+
+    h.value = True
+    h._tool.click(0.0, -30.0)
+    h._tool.click(0.0, 1.0)
+
+    file = tempfile.NamedTemporaryFile(suffix=".json")
+    masking_tool.filename.value = file.name
```
Comment (jl-wynen) [inline tests/masking_tool_test.py:226]: Why not provide the filename as an argument to `save_masks`? This here seems like a strange interface to me.

File: tests/masking_tool_test.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    assert mask['y']["min"] == {"value": -30.0, "unit": "m"}
+    assert mask['y']["max"] == {"value": 1.0, "unit": "m"}
+
+
+@pytest.mark.usefixtures('_use_ipympl')
+def test_trying_to_save_masks_with_no_masks_raises():
+    da = make_data()
+    masking_tool = MaskingTool(da)
+    masking_tool.filename.value = "masks.json"
+    with pytest.raises(ValueError, match="There are no masks to save."):
```
Comment (jl-wynen) [inline tests/masking_tool_test.py:259]: Should it really work this way? What is the harm in saving an empty file?

File: tests/masking_tool_test.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    h.value = True
+    h._tool.click(0.0, -30.0)
+    h._tool.click(0.0, 1.0)
+
+    file = tempfile.NamedTemporaryFile(suffix=".json")
+    masking_tool.filename.value = file.name
+    masking_tool.save_masks()
+
+    with open(file.name) as f:
+        loaded_masks = json.load(f)
```
Comment (jl-wynen) [inline tests/masking_tool_test.py:230]: Are you planning to implement a `load_masks` function?

File: src/scippneutron/masking.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+        """
+        import ipywidgets as ipw
+        from mpltoolbox import Hspans, Rectangles, Vspans
+        from plopp.widgets import DrawingTool, style
+
+        # Convert potential bin edge coords to midpoints
+        da = data.copy(deep=False)
+        for dim, coord in data.coords.items():
+            if data.coords.is_edges(dim):
+                da.coords[dim] = sc.midpoints(coord)
```
Comment (jl-wynen) [inline src/scippneutron/masking.py:125]:
```suggestion
        for name, coord in data.coords.items():
            if data.coords.is_edges(name):
                da.coords[name] = sc.midpoints(coord)
```

File: src/scippneutron/masking.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+        ndim = da.ndim
+        self.data_node = pp.Node(da)
+        self.masking_node = pp.Node(_apply_masks, self.data_node)
+        self.fig = {1: pp.linefigure, 2: pp.imagefigure}[ndim](
```
Comment (jl-wynen) [inline src/scippneutron/masking.py:130]: This would lead to a confusing error message for data with bad `ndim`. Can you catch that case explicitly?

File: src/scippneutron/masking.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+        if change["new"]:
+            for c in self.controls:
+                if c.value and c is not change["owner"]:
+                    c.value = False
+
+    def validate_filename(self, change: dict) -> None:
+        path = change["new"] + ".json"
+        if os.path.exists(path):
+            self.filename.style.background = "#ff9999"
+            self.save_button.disabled = True
```
Comment (jl-wynen) [inline src/scippneutron/masking.py:210]: Ok for a first version. But I'm sure that people will complain.

File: src/scippneutron/masking.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    """
+    x = rect_info["x"]
+    y = rect_info["y"]
+    xcoord = da.coords[x["dim"]]
+    ycoord = da.coords[y["dim"]]
+    return (
+        (xcoord >= x["min"])
+        & (xcoord <= x["max"])
+        & (ycoord >= y["min"])
+        & (ycoord <= y["max"])
```
Comment (jl-wynen) [inline src/scippneutron/masking.py:32]:
```suggestion
    return (
        ((xcoord >= x["min"])
        & (xcoord <= x["max"]))
        & ((ycoord >= y["min"])
        & (ycoord <= y["max"]))
```
I think this way, it does two operations on 1d arrays and one op on a 2d array. With you version, it is one op on 1d and two ops on 2d.

File: src/scippneutron/masking.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        "y": {
+            "dim": figure.canvas.dims["y"],
+            "min": sc.scalar(min(y1, y2), unit=figure.canvas.units["y"]),
+            "max": sc.scalar(max(y1, y2), unit=figure.canvas.units["y"]),
+        },
+    }
+
+
+def _define_span_mask(da: sc.DataArray, span_info: dict) -> sc.Variable:
+    info = next(iter(span_info.values()))
```
Comment (jl-wynen) [inline src/scippneutron/masking.py:60]: What about the other elements in `span_info`?

File: docs/user-guide/masking-tool.ipynb [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "812ac2e0-dcb9-4405-931a-5e2dc77db9bd",
+   "metadata": {},
+   "source": [
+    "## Two-dimensional masks\n",
+    "\n",
+    "We then load it into the `MaskingTool`:"
```
Comment (jl-wynen) [inline docs/user-guide/masking-tool.ipynb:62]: Can you add an explanation how the masks in this example are defined? It may be surprising for users when they try to follow the explanation but end up without any masks. 
It may even be best to show the code the creates the masks. See also the comment about making the programmatic interface public.

File: docs/user-guide/masking-tool.ipynb [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+    "v._tool.click(7, 0)\n",
+    "v._tool.click(15, 0)\n",
+    "v._tool.click(20, 0)\n",
+    "v._tool.click(30, 0)\n",
+    "\n",
+    "h.value = True\n",
+    "h._tool.click(0, -20)\n",
+    "h._tool.click(0, -10)\n",
+    "\n",
+    "masking_tool.filename.value = \"masks-2d.h5\""
```
Comment (jl-wynen) [inline docs/user-guide/masking-tool.ipynb:100]:
```suggestion
    "masking_tool.filename.value = \"masks-2d.json\""
```

File: docs/user-guide/masking-tool.ipynb [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+   "outputs": [],
+   "source": [
+    "v = masking_tool.controls[1]\n",
+    "v.value = True\n",
+    "v._tool.click(-40, 150)\n",
+    "v._tool.click(-20, 150)\n",
+    "v._tool.click(20, 150)\n",
+    "v._tool.click(40, 150)\n",
+    "\n",
+    "masking_tool.filename.value = \"masks-1d.h5\""
```
Comment (jl-wynen) [inline docs/user-guide/masking-tool.ipynb:169]:
```suggestion
    "masking_tool.filename.value = \"masks-1d.json\""
```

File: docs/user-guide/masking-tool.ipynb [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1eead265-50ad-46ea-ad43-ef90dff6be81",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "masking_tool"
```
Comment (jl-wynen) [inline docs/user-guide/masking-tool.ipynb:110]: Can you add an explanation for how to manipulate masks? I found myself clicking around a lot and accidentally creating a lot of masks.

File: docs/user-guide/masking-tool.ipynb
Comment (jl-wynen) [inline docs/user-guide/masking-tool.ipynb:1]: Is it possible to remove masks? It would not be great to have to start over every time you accidentally add a new mask (easy by misclicking).

Also, an undo button would go a long way. But that might be difficult and a task for another PR.

File: src/scippneutron/masking.py [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+        )
+        hspans = DrawingTool(
+            tool=partial(Hspans, **col_args),
+            get_artist_info=_get_hspan_info,
+            icon="grip-lines",
+            func=_define_span_mask,
+            disabled=ndim == 1,
+            **common,
+        )
+        self.controls = [rects, vspans, hspans]
```
Comment (jl-wynen) [inline src/scippneutron/masking.py:165]: These tools don't have tooltips.

**Review by jl-wynen — COMMENTED** [review]
I have a lot of comments but they are mostly about small issues. It looks good overall!

File: pyproject.toml (hunk 1 above)
Comment (nvaytet) [reply pyproject.toml:91]: This is also [in Plopp](https://github.com/scipp/plopp/blob/main/pyproject.toml#L62). The warning pops up when using `ipympl`. I was never able to fix it.

File: tests/conftest.py (hunk 3 above)
Comment (nvaytet) [reply tests/conftest.py:22]: hmm I thought it was using `ipympl` for only a given test, but maybe it sets it for all?

File: requirements/basetest.in (hunk 2 above)
Comment (nvaytet) [reply requirements/basetest.in:5]: I tried this now but I messes things up in the nightly dependencies, where we want to get `plopp` and `mpltoolbox` from `@main`... :-(

File: tests/masking_tool_test.py (hunk 5 above)
Comment (nvaytet) [reply tests/masking_tool_test.py:226]: Yes, I'll do that.

File: tests/masking_tool_test.py (hunk 6 above)
Comment (nvaytet) [reply tests/masking_tool_test.py:259]: Don't know, wanted to prevent the case where a user would think they are saving something but in fact they are not?
I don't mind either way.

File: tests/masking_tool_test.py (hunk 7 above)
Comment (nvaytet) [reply tests/masking_tool_test.py:230]: This would just be `json.load()` so I thought we didn't really need it?
Also, where would you put it? I don't think it makes much sense to have it as a method on the tool?..

Ah, unless you want to load some old masks in the tool and edit them further?

Comment (jl-wynen) [reply tests/masking_tool_test.py:230]: > Ah, unless you want to load some old masks in the tool and edit them further?

Exactly. This could also be used to inspect masks.

File: src/scippneutron/masking.py (hunk 10 above)
Comment (jl-wynen) [reply src/scippneutron/masking.py:210]: :+1: for the current solution for now. I think in the long run a dialogue would be best. But I don't know ow difficult that is in ipywidgets.

File: src/scippneutron/masking.py (hunk 12 above)
Comment (nvaytet) [reply src/scippneutron/masking.py:60]: I think there can only be one entry in that dict?
I used this because the key can be either `x` or `y`, depending on the span direction, but I know there is only one value.
The info is generated by `_get_vspan_info` and `_get_hspan_info`.

File: docs/user-guide/masking-tool.ipynb (hunk 16 above)
Comment (nvaytet) [reply docs/user-guide/masking-tool.ipynb:110]: I added a help button that has a tooltip with instructions.
I unfortunately could not add just an icon with a tooltip, it had to be a button, but clicking the button does nothing (see https://github.com/jupyter-widgets/ipywidgets/issues/3703#issuecomment-2154581519).

I experimented with having hidden instructions that would appear if you clicked the button but in the end, this was the simplest and didn't mess up the layout of the figure.

File: docs/user-guide/masking-tool.ipynb
Comment (nvaytet) [reply docs/user-guide/masking-tool.ipynb:1]: > Is it possible to remove masks?

You can remove them by middle-clicking. I added this to the instructions.

> Also, an undo button would go a long way. But that might be difficult and a task for another PR.

Yes, difficult and annoying. We'd have to keep a history etc...

File: .copier-answers.yml [hunk 18]
```diff
@@ -5,7 +5,7 @@ description: Neutron scattering tools for Data Reduction
 max_python: '3.12'
 min_python: '3.10'
 namespace_package: ''
-nightly_deps: scipp,scippnexus,plopp
+nightly_deps: scipp,scippnexus,plopp,mpltoolbox
```
Comment (nvaytet) [inline .copier-answers.yml:8]: I had to edit this manually because it was needed for the nightly deps and I didn't want to do a copier update at the same time

File: src/scippneutron/masking.py (hunk 10 above)
Comment (nvaytet) [reply src/scippneutron/masking.py]: I find confirmation dialogs also annoying, but I guess they provide some safety...

File: tests/masking_tool_test.py (hunk 7 above)
Comment (nvaytet) [reply tests/masking_tool_test.py:228]: We would have to check that the data that has been loaded is compatible with the masks in the file (axes dimensions and units match etc.). I would leave that to a later PR I think.

File: tests/conftest.py (hunk 3 above)
Comment (nvaytet) [reply tests/conftest.py:22]: Ok it is indeed setting it for all the other tests

Comment (nvaytet) [reply tests/conftest.py:26]: I think I managed to fix this by switching to ipympl for the test and switching back to matplotlib defaults after the test.

File: src/scippneutron/masking.py (hunk 12 above)
Comment (jl-wynen) [reply src/scippneutron/masking.py:57]: Can you use a tuple of the form `(dim, span)` instead?

File: src/scippneutron/masking.py [hunk 19]
```diff
@@ -105,6 +105,15 @@ def __init__(self, data: sc.DataArray, color='magenta', **kwargs):
...
         buttons in the top bar of the figure.
 
+        Instructions:
+
+        • Use the buttons in the top bar to add masks to the data
+        • Left-click to add a new shape, and left-click again to persist the shape
+        • Left-click a vertex to edit a shape
+        • Right-click and hold to drag a shape
+        • Middle-click (or Ctrl + left-click) to delete a shape
+        • Save the masks to a file when the "Save" button is clicked
```
Comment (jl-wynen) [inline src/scippneutron/masking.py:112]: What are these bullet characters? Are they rendered properly by Sphinx?

Comment (nvaytet) [reply src/scippneutron/masking.py]: I made bullets in the tooltip, and copy/pasted. I'll change to `-`.

**Comment by jl-wynen** [comment]
One last comment that may be hard to find in the list: https://github.com/scipp/scippneutron/pull/521#discussion_r1634408877

File: src/scippneutron/masking.py (hunk 12 above)
Comment (nvaytet) [reply src/scippneutron/masking.py:57]: I tried, or simply returning a dict without the initial `x` or `y` key, but it then fails when I call `get_masks` because that expects a standard format for all artist infos. But maybe my infos are bad?
Maybe they should also contain the kind of artists, e.g. `rectangle` and `vspan`, so it's easier to know what to do when we load them again?

**Review by jl-wynen — APPROVED** [review]
